presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -216,6 +217,159 @@ else if (rescaleFactor > 0) {
+            // either both negative or both positive
+            overflow = addUnsignedReturnOverflow(left, right, result, leftNegative);
+            if (leftNegative) {
+                overflow = -overflow;
+            }
+        }
+        else {
+            int compare = compareAbsolute(left, right);
+            if (compare > 0) {
+                subtractUnsigned(left, right, result, leftNegative);
+            }
+            else if (compare < 0) {
+                subtractUnsigned(right, left, result, !leftNegative);
+            }
+            else {
+                clearDecimal(result);
+            }
+        }
+        return overflow;
+    }
+
+    public static Slice subtract(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+        subtract(left, right, result);
+        return result;
+    }
+
+    public static void subtract(Slice left, Slice right, Slice result)
+    {
+        if (isNegative(left) ^ isNegative(right)) {
+            // only one is negative
+            if (addUnsignedReturnOverflow(left, right, result, isNegative(left)) != 0) {
+                throwOverflowException();
+            }
+        }
+        else {
+            int compare = compareAbsolute(left, right);
+            if (compare > 0) {
+                subtractUnsigned(left, right, result, isNegative(left) && isNegative(right));
+            }
+            else if (compare < 0) {
+                subtractUnsigned(right, left, result, !(isNegative(left) && isNegative(right)));
+            }
+            else {
+                clearDecimal(result);
+            }
+        }
+    }
+
+    /**
+     * This method ignores signs of the left and right. Returns overflow value.
+     */
+    private static long addUnsignedReturnOverflow(Slice left, Slice right, Slice result, boolean resultNegative)
+    {
+        int l0 = getInt(left, 0);
+        int l1 = getInt(left, 1);
+        int l2 = getInt(left, 2);
+        int l3 = getInt(left, 3);
+
+        int r0 = getInt(right, 0);
+        int r1 = getInt(right, 1);
+        int r2 = getInt(right, 2);
+        int r3 = getInt(right, 3);
+
+        long product;
+        product = (l0 & LONG_MASK) + (r0 & LONG_MASK);
+
+        int z0 = (int) product;
+
+        product = (l1 & LONG_MASK) + (r1 & LONG_MASK) + (product >> 32);
+
+        int z1 = (int) product;
+
+        product = (l2 & LONG_MASK) + (r2 & LONG_MASK) + (product >> 32);
+
+        int z2 = (int) product;
+
+        product = (l3 & LONG_MASK) + (r3 & LONG_MASK) + (product >> 32);
+
+        int z3 = (int) product & (~SIGN_INT_MASK);
+
+        pack(result, z0, z1, z2, z3, resultNegative);
+
+        return product >> 31;
+    }
+
+    /**
+     * This method ignores signs of the left and right and assumes that left is greater then right
+     */
+    private static void subtractUnsigned(Slice left, Slice right, Slice result, boolean resultNegative)
+    {
+        int l0 = getInt(left, 0);
```

Comment:
One idea for possible future optimization is to do 2 7-byte operations (by reading a long an masking the top byte to leave space for carry) and one 2-byte operation instead of 4 4-byte operations. It may not matter, but it might be worth benchmarking at some point.